pkg/operator/ceph/cluster/osd/deviceset_test.go

```diff
@@ -83,10 +80,105 @@ func testPrepareDeviceSets(t *testing.T, setTemplateName bool) {
+	assert.NoError(t, err)
+	assert.Equal(t, 6, len(pvcs.Items))
+
+	// Verify the same number of PVCs exist after calling the reconcile again on the PVCs
+	volumeSources = cluster.prepareStorageClassDeviceSets(config)
+	assert.Equal(t, 2, len(volumeSources))
+	pvcs, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).List(ctx, metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 6, len(pvcs.Items))
+
+	// Delete a single PVC and verify it will be re-created
+	err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Delete(ctx, "rook-ceph-osd-mydata-wal-0", metav1.DeleteOptions{})
+	assert.NoError(t, err)
+	volumeSources = cluster.prepareStorageClassDeviceSets(config)
+	assert.Equal(t, 2, len(volumeSources))
+	pvcs, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).List(ctx, metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 6, len(pvcs.Items))
+
+	// Delete the PVCs for an OSD and verify it will not be re-created if the count is reduced
+	cluster.spec.Storage.StorageClassDeviceSets[0].Count = 1
+	err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Delete(ctx, "rook-ceph-osd-mydata-data-0", metav1.DeleteOptions{})
+	assert.NoError(t, err)
+	err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Delete(ctx, "rook-ceph-osd-mydata-metadata-0", metav1.DeleteOptions{})
+	assert.NoError(t, err)
+	err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Delete(ctx, "rook-ceph-osd-mydata-wal-0", metav1.DeleteOptions{})
+	assert.NoError(t, err)
+	volumeSources = cluster.prepareStorageClassDeviceSets(config)
+	assert.Equal(t, 1, len(volumeSources))
+	pvcs, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).List(ctx, metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 3, len(pvcs.Items))
+
+	// Scale back up to a count of two and confirm that a new index is used for the PVCs
+	cluster.spec.Storage.StorageClassDeviceSets[0].Count = 2
+	volumeSources = cluster.prepareStorageClassDeviceSets(config)
+	assert.Equal(t, 2, len(volumeSources))
+	pvcs, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).List(ctx, metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 6, len(pvcs.Items))
+	pvc, err := clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Get(ctx, "rook-ceph-osd-mydata-data-2", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, pvc)
+	pvc, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Get(ctx, "rook-ceph-osd-mydata-metadata-2", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, pvc)
+	pvc, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Get(ctx, "rook-ceph-osd-mydata-wal-2", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, pvc)
+}
```

Comment:
Making two tests are better. One is for making a hole case that removes OSD0 (it already exists). The other is for the non-hole case (it doesn't exist yet).

Reply:
I think those test cases are already covered, right? Are you just suggesting to split the new test into two functions? I'd prefer to keep it as one unless it's a different test I missed.

Reply:
I think the suggestion is to add a second test that ensures that if the highest-indexed ID is deleted, it gets reused when scaled back up. This might be a new test like `TestPrepareDeviceSetWithoutHolesInPVC`

Reply:
I think that this is implicit already in the tests where it verified that the first available index (2) was used to add a new osd. If there are no holes in the index, it's just the base case compared to if there were holes.

Reply:
I do agree that it's not necessary, but I think it's still valuable to add that expectation to the test. In some respect, it is testing the corner case of _remove and re-create the **last** list item_ just as the first test does the same for the _**first** item_ where first and last items in a list are often corner cases. If I were to write the tests, I would probably test behavior with zero OSDs present, with one OSD present, and all 3 cases of removing and re-creating one item from a 3-item list to be sure there weren't unexpected logic issues.

Also, I have been advised before that the surest way to determine what code does is to examine the tests.

Reply:
I'll add one more step to the tests that will start with count of 1, then increase to 2, rather than just starting with 2. I believe this will cover the test hole you're looking for.